drivers/include/periph/qspi.h

```diff
@@ -0,0 +1,267 @@
+ */
+#ifndef QSPI_DEV
+#define QSPI_DEV(x)     (x)
+#endif
+
+/**
+ * @brief   Define global value for undefined QSPI device
+ */
+#ifndef QSPI_UNDEF
+#define QSPI_UNDEF      (UINT_MAX)
+#endif
+
+/**
+ * @brief   Default type for QSPI devices
+ */
+#ifndef HAVE_QSPI_T
+typedef unsigned int qspi_t;
+#endif
+
+/**
+ * @brief   Available QSPI modes, defining the configuration of clock polarity
+ *          and clock phase
+ *
+ * RIOT is using the mode numbers as commonly defined by most vendors
+ * (https://en.wikipedia.org/wiki/Serial_Peripheral_Interface_Bus#Mode_numbers):
+ *
+ * - MODE_0: CPOL=0, CPHA=0 - The first data bit is sampled by the receiver on
+ *           the first SCK rising SCK edge (this mode is used most often).
+ * - MODE_1: CPOL=0, CPHA=1 - The first data bit is sampled by the receiver on
+ *           the second rising SCK edge.
+ * - MODE_2: CPOL=1, CPHA=0 - The first data bit is sampled by the receiver on
+ *           the first falling SCK edge.
+ * - MODE_3: CPOL=1, CPHA=1 - The first data bit is sampled by the receiver on
+ *           the second falling SCK edge.
+ */
+#ifndef HAVE_QSPI_MODE_T
+typedef enum {
+    QSPI_MODE_0 = 0,        /**< CPOL=0, CPHA=0 */
+    QSPI_MODE_1,            /**< CPOL=0, CPHA=1 */
+    QSPI_MODE_2,            /**< CPOL=1, CPHA=0 */
+    QSPI_MODE_3             /**< CPOL=1, CPHA=1 */
+} qspi_mode_t;
+#endif
+
+/**
+ * @brief   Standard QSPI erase block sizes
+ */
+#ifndef HAVE_QSPI_ERASE_SIZE_T
+typedef enum {
+    QSPI_ERASE_4K   = SFLASH_CMD_ERASE_SECTOR,
+    QSPI_ERASE_64K  = SFLASH_CMD_ERASE_BLOCK,
+    QSPI_ERASE_CHIP = SFLASH_CMD_ERASE_CHIP,
+} qspi_erase_size_t;
+#endif
+
+/**
+ * @brief   Configuration Options for QSPI
+ * @{
+ */
+#define QSPI_FLAG_1BIT          (0x0)   /**< single data line           */
+#define QSPI_FLAG_2BIT          (0x1)   /**< two parallel data lines    */
+#define QSPI_FLAG_4BIT          (0x2)   /**< four parallel data lines   */
+#define QSPI_FLAG_8BIT          (0x3)   /**< eight parallel data lines  */
+
+#define QSPI_FLAG_BIT_MASK      (0x3)   /**< mask to get data lines from flags */
+#define QSPI_FLAG_BIT(flags)    ((flags) & QSPI_FLAG_BIT_MASK)
+
+#define QSPI_FLAG_DDR           (0x4)   /**< use Double Data Rate mode  */
+
+#define QSPI_FLAG_ADDR_32BIT    (0x8)   /**< use 32 bit addressing      */
+/** @} */
+
+/**
+ * @brief   Basic initialization of the given QSPI bus
+ *
+ * @note    This function MUST not be called more than once per bus!
+ *
+ * @param[in] bus       QSPI device to initialize
+ */
+void qspi_init(qspi_t bus);
+
+/**
+ * @brief   Configure the QSPI peripheral settings
+ *
+ * @pre     The @p bus has been acquired with @ref qspi_acquire
+ *
+ * @param[in] bus       QSPI device to configure
+ * @param[in] mode      QSPI mode @see qspi_mode_t
+ * @param[in] flags     QSPI Configuration Options
+ * @param[in] clk_hz    QSPI frequency in Hz
+ */
+void qspi_configure(qspi_t bus, qspi_mode_t mode, uint32_t flags, uint32_t clk_hz);
```

Comment:
qn9080 and nrf52840 both support memory mapping a flash to a region. To do that you need to program a command (like the fast_read command) and an address format, so then you can read from MCU memory directly which will generate a QSPI transaction automatically, even with some caching, allowing very fast reads without any call overhead. This is often incompatible with running a command at the same time, so something like qspi_configure but for configuring memory reads that returns the pointer/size would be good to add here. It is probably an addition to this module, meaning that you would need to have MODULE_QSPI_MAPPING defined, but it is worth considering the idea in this design so we don't have to change it later.

Reply:
SAM D5x/E5x has that too, but the memory region where this is mapped to is fixed.
I have a [branch](https://github.com/benpicco/RIOT/tree/cpu/same54-qspi-xip) where I implemented `qspi_xip_mount()` - haven't pushed that yet because I couldn't get the flash chip out of XIP mode without doing a power reset.

I assumed that in this mode you can only read, not write to the flash.

Reply:
Reading the datasheet it seems that you can only read, but it is not very clear. The nRF52840 has registers for configuring the READ, WRITE and ERASE opcodes, but I can't quite tell if the chip handles writes or not. QN9080 says it only handles reads and actually XIP is not recommended because it is too slow.

I think this is another example of code that shouldn't go into the qspi cpu driver: sending commands to enabled/disable the xip mode in the flash chip (looking at your code I understand you mean no-opcode mode). Some chips use 0xa5 for entering this mode and 0xff to exit, but this depends on the flash chip, not the cpu, and it varies a bit across manufacturers. For this what's needed is a way to set the dummy bytes to send (probably in the flags, or similar).

The XIP mount memory region is normally a window of certain fixed size and address; but you could have a different base flash address if the flash is larger than the window. I don't see this feature in the samd5x, but nrf52840 has the XIPOFFSET register for this purpose. I think that when configuring the memory mapping (`xip_mount` in your branch) the caller should pass the configuration: command (if any), whether an command is needed, the flash base address; and the driver would return the memory address corresponding to the requested base address. If windowing is supported you would set that in the xipoffset register for example, but if there are alignment requirements for this offset or having an offset is not supported the driver could return an address into the fixed region (like `AHB_XIP+base_address` or `AHB_XIP + base_address % min_alignment` while setting the offset to another register). The caller in this example would be a qspi_nor driver, which is different from a qspi_flash driver for example but the same across CPUs. I'd imagine that the qspi_nor driver can send commands to enable the XIP mode in the flash (depending on the flash) and then tell the qspi.h interface to do xip_mount with no opcode; or don't set the XIP mode in the flash if not supported and then tell the qspi.h to do xip_mount using certain opcode for reads.